rfcs/0159-private-task-logs.md

```diff
@@ -0,0 +1,88 @@
+# RFC 159 - Private Task Logs
+* Comments: [Bug 1598698](https://bugzilla.mozilla.org/show_bug.cgi?id=1598689), [#159](https://github.com/taskcluster/taskcluster-rfcs/pull/159)
+* Proposed by: @djmitche
+
+# Summary
+
+A task's log output is currently world-readable.
+Tasks that wish to keep their output private must carefully divert output from stdout/stderr into other files and upload those files as artifacts.
+
+This RFC proposes making this behavior optional, with private logs as an alternative.
+Such logs would require credentials with specific scopes to read.
+
+## Motivation
+
+Even in open-source development, secrecy can be important.
+For example, CI for fixes to security issues can reveal those security issues if logs are public.
+
+Private-log capability is also a common request from outside parties interested in deploying Taskcluster.
+
+# Details
+
+## Current Log Implementation
+
+The current [convention](https://docs.taskcluster.net/docs/manual/design/conventions/task-logs) is to name the main task log `public/logs/live.log`, and that fetching this artifact may -- if the worker supports it -- result in receiving log lines as they are produced.
+
+The Taskcluster UI, and other consumers of task logs such as Treeherder, incorporate this convention and look for logs at that artifact name.
+
+Workers that cannot do "live logging" simply create an S3 artifact with this name.
+Workers that *can* do live logging use the following process.
+At the beginning of the task, they create a Reference artifact named `public/logs/live.log` containing a URL that will connect to the worker (usually via [Websocktunnel](https://github.com/taskcluster/taskcluster/tree/master/tools/websocktunnel#readme)) to get log lines as they are produced.
+When the task is complete, they create a new S3 artifact containing the whole log, named `public/logs/live_backing.log`.
+They then modify the URL in the `public/logs/live.log` artifact (using a [special case in the `createTask` API](https://docs.taskcluster.net/docs/reference/platform/queue/api#createArtifact)) to point to the URL for the `getArtifact` method for the `public/logs/live_backing.log` artifact (`<rootUrl>/api/queue/v1/task/<taskId>/runs/<runId>/artifacts/public/logs/live_backing.log`).
+
+Reference artifacts correspond to HTTP 303 redirects, and HTTP clients do not re-authenticate when following a redirect.
```

Comment:
Why is a redirect not able to re-authenticate? Are headers omitted on a redirect?

Reply:
> Are headers omitted on a redirect?

[Yes](https://stackoverflow.com/a/28671822/2737366)

Reply:
That makes a lot of sense. Thank you.

Reply:
Oops, I meant to add that to the RFC -- I'll do that now.